GEWIS/gewisweb
# Upgraded the project to Laminas Framework

The Frontpage is loading already but we should verify a couple of things and preferably refactor a bit to make the project a bit more maintainable in the future.
While refactoring to upgrade to ZF3 I encountered quite some architectural issues.

With this update we get: 
- A more up-to-date version of our framework. We skip ZF3 and go directly from ZF2 to Laminas Framework.
- A more up-t-date version of PHP. We now have PHP version 7.4 with high chances of being able to upgrade to PHP 8.0 in the future
- Less outdated dependency warnings
- Less abandoned packages warnings
- PSR-4 file structure in modules
- Composer autoloading instead of zend-loader or laminas-loader
- A renewed ACL interface with seperate services for each module. Less code duplication and better seperation of services.
- A more secure website probably
- A faster website?
- Did I say less warnings?

TODO before merging:
- Re-enable the ApiUser role in the `user-role` service after resolving circular dependencies
- See if admin role permissions work correctly
- Review TODO marks I left within the code and resolve where neccessary (note that some relate to pre-existing problems)
- Review effects on ORM patches
- Do manual testing to see if everything works
- Resolve merge conflicts

TODO after merging (will probably create seperate issues for these later):
- Create dedicated services within each module that expose relevant functionality to other modules (facade design pattern). This helps prevent circular dependencies between modules
- God damn, improve that code quality
- Move ACL checks from services to the controller?
- More refactoring?
- Apply updates to individual components (check `php composer.phar outdated`)
- Currently glide-zend still has a dependency on zend-diactoros. This is not blocking functionality but it is blocking upgrades in the future. We should probably drop this dependency at some point. (https://github.com/thephpleague/glide-zend/issues/2)
- Add return types to methods and parameter types as well. These are type-checked contrary to the PhpDoc comments and will thus improve the reliability of our code and debugging experience.
